### PR TITLE
Add an intrinsic for Hash#to_h

### PR DESCRIPTION
**Pull request by elliottt**
Adds an intrinsic for `Hash#to_h`.

### Motivation

Performance.

### Test plan

See included automated tests.

### CODE REVIEW

File: test/testdata/compiler/intrinsics/identities.rb [hunk 1]
```diff
@@ -73,6 +93,15 @@ def self.array_to_a(x)
...
 p M.hash_to_hash({x: 'hi', 10 => false})
 
+class HashSubclass < Hash
+end
+
+p M.hash_to_h({x: 'hi', 10 => false})
+p M.hash_to_h(HashSubclass.new)
+
+p M.hash_to_h_withBlock({x: 'hi', 10 => false})
+p M.hash_to_h_withBlock(HashSubclass.new)
```
Comment (aprocter-stripe) [inline test/testdata/compiler/intrinsics/identities.rb:103]:
```suggestion
p M.hash_to_h({x: 'hi', 10 => false})
p M.hash_to_h({x: 'hi', 10 => false}).class
p M.hash_to_h(HashSubclass.new)
p M.hash_to_h(HashSubclass.new).class

p M.hash_to_h_withBlock({x: 'hi', 10 => false})
p M.hash_to_h_withBlock({x: 'hi', 10 => false}).class
p M.hash_to_h_withBlock(HashSubclass.new)
p M.hash_to_h_withBlock(HashSubclass.new).class
```

File: compiler/IREmitter/Payload/codegen-payload.c [hunk 2]
```diff
@@ -685,6 +685,50 @@ VALUE sorbet_arrayNewFromValues(int argc, VALUE *argv) {
...
+SORBET_INLINE
+VALUE sorbet_int_hash_to_h(VALUE recv, ID fun, int argc, const VALUE *const restrict argv, BlockFFIType blk,
+                           VALUE closure) {
+    extern VALUE sorbet_rb_hash_to_h(VALUE);
+    return sorbet_rb_hash_to_h(recv);
+}
+
+struct sorbet_int_hash_to_h_closure {
+    BlockFFIType fun;
+    VALUE toplevel_closure;
```
Comment (aprocter-stripe) [inline compiler/IREmitter/Payload/codegen-payload.c:700]: Not-very-strongly-felt suggestion: maybe `closure_for_block` would be a better name?

**Review by aprocter-stripe — APPROVED** [review]
LGTM! Just a couple minor suggestions.

File: compiler/IREmitter/Payload/codegen-payload.c (hunk 2 above)
Comment (elliottt) [reply compiler/IREmitter/Payload/codegen-payload.c:700]: Works for me!

**Comment by elliottt** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_KcsDwTBCkj7wse
→ https://go/builds/bui_KcsDFVKh30qRdV

**Pull request merged**